File: debug_divergent_measurements.py
#!/usr/bin/env python3
"""
Debug script to trace processing of divergent measurements.
Compares TypeScript and Python behavior for specific measurement IDs.
"""

import csv
import sys
from pathlib import Path
from datetime import datetime, timezone

# Add paths for imports
project_root = Path(__file__).parent
sys.path.insert(0, str(project_root / "python_lib" / "src"))
sys.path.insert(0, str(project_root / "be_implementation_service" / "src"))

from aws.api.models import Measurement
from aws.services.weight_processor_service import WeightProcessorService
from aws.config.config_manager import ConfigManager
from weight_processor_lib.core.database import InMemoryStore

# Target measurement IDs that differ between implementations
DIVERGENT_IDS = [
    "726b441f-eb43-47d9-8f3c-845d164e5a5b",  # TS accepts, PY rejects
    "1a98b2c3-e023-4757-8d01-d35ef2fb363e",  # PY accepts, TS rejects
    "510977fa-9d3f-4b50-a667-e676a0cc0791",  # PY accepts, TS rejects
    "70d7918e-87d9-4968-84a4-b2bfec488e76",  # PY accepts, TS rejects
    "86233705-0332-44f1-bc69-bc796220f598",  # PY accepts, TS rejects
]

def parse_timestamp(date_str: str) -> datetime:
    """Parse various timestamp formats and return timezone-aware datetime in UTC."""
    if not date_str:
        return datetime.now(timezone.utc)

    try:
        if "T" in date_str:
            dt = datetime.fromisoformat(date_str.replace("Z", "+00:00"))
            if dt.tzinfo is None:
                return dt.replace(tzinfo=timezone.utc)
            return dt.astimezone(timezone.utc)
        elif " " in date_str:
            for fmt in ["%Y-%m-%d %H:%M:%S.%f", "%Y-%m-%d %H:%M:%S"]:
                try:
                    dt = datetime.strptime(date_str, fmt)
                    return dt.replace(tzinfo=timezone.utc)
                except ValueError:
                    continue
            raise ValueError(f"Cannot parse space-separated date: {date_str}")
        else:
            dt = datetime.strptime(date_str, "%Y-%m-%d")
            return dt.replace(tzinfo=timezone.utc)
    except Exception:
        return datetime.now(timezone.utc)

def main():
    print("=== Debugging Divergent Measurements (Python) ===\n")
    print("Target IDs:")
    for id_ in DIVERGENT_IDS:
        print(f"  {id_}")
    print()

    # Load CSV
    csv_path = "test_user.csv"
    user_measurements = []
    divergent_measurements = []

    with open(csv_path, 'r') as f:
        reader = csv.DictReader(f)
        for row in reader:
            measurement_id = row.get("id") or row.get("measurement_id")
            user_id = row.get("user_id")
            weight_str = (row.get("value_quantity", "") or row.get("weight", "")).strip()

            if not user_id or not measurement_id or not weight_str:
                continue

            try:
                weight = float(weight_str)
                if weight <= 0 or weight > 1000:
                    continue
            except (ValueError, TypeError):
                continue

            date_str = row.get("effective_date_time", "") or row.get("effectiveDateTime", "")
            source = row.get("source_type", "unknown")
            unit = row.get("unit", "").strip()

            if not unit:
                continue

            timestamp = parse_timestamp(date_str) if date_str else datetime.now(timezone.utc)

            measurement = Measurement(
                uuid=measurement_id,
                weight=weight,
                unit=unit,
                effectiveDateTime=timestamp,
                source=source,
                metadata={"original_row": row}
            )

            user_measurements.append(measurement)

            if measurement_id in DIVERGENT_IDS:
                divergent_measurements.append((measurement, row))

    print(f"Loaded {len(user_measurements)} measurements for user")
    print(f"Found {len(divergent_measurements)} divergent measurements\n")

    # Sort by timestamp
    user_measurements.sort(key=lambda m: m.measured_at)

    # Initialize Python implementation
    config = ConfigManager.load_config(source="file")
    config["database"]["backend"] = "memory"
    store = InMemoryStore()
    service = WeightProcessorService(state_store=store, config=config)

    user_id = "ADC64C0B-CB46-41F9-BDA0-CC11A35942D7"

    # Process all measurements
    print("Processing all measurements...\n")
    response = service.process_batch(user_id, user_measurements)

    print(f"Total processed: {response.measurements_processed}")
    print(f"Total accepted: {response.measurements_accepted}")
    print(f"Total rejected: {response.measurements_rejected}\n")

    # Analyze divergent measurements
    print("=== Divergent Measurement Analysis ===\n")

    for measurement, original_row in divergent_measurements:
        # Find the result for this measurement
        measurement_index = next(
            (i for i, m in enumerate(user_measurements) if m.measurement_id == measurement.measurement_id),
            -1
        )

        if 0 <= measurement_index < len(response.results):
            result = response.results[measurement_index]

            print(f"ID: {measurement.measurement_id}")
            print(f"  Timestamp: {measurement.measured_at.isoformat()}")
            print(f"  Source: {measurement.source}")
            print(f"  Weight: {measurement.weight_value} {measurement.weight_unit}")
            print(f"  Accepted: {result.accepted}")
            print(f"  Quality Score: {result.quality_score:.6f}" if result.quality_score else "  Quality Score: None")

            if result.quality_components:
                print(f"  Quality Components:")
                for component, score in result.quality_components.items():
                    print(f"    {component}: {score:.6f}")

            if not result.accepted:
                print(f"  Rejection Reason: {result.rejection_reason or 'Unknown'}")

            print(f"  Kalman Estimate: {result.kalman_estimate:.3f}" if result.kalman_estimate else "  Kalman Estimate: None")
            print(f"  Innovation: {result.innovation:.3f}" if result.innovation is not None else "  Innovation: None")
            print(f"  Normalized Innovation: {result.normalized_innovation:.3f}" if result.normalized_innovation is not None else "  Normalized Innovation: None")
            print()

if __name__ == "__main__":
    main()

File: debug_divergent_measurements.ts
#!/usr/bin/env bun
/**
 * Debug script to trace processing of divergent measurements.
 * Compares TypeScript and Python behavior for specific measurement IDs.
 */

import { readFileSync } from "fs";
import { parse } from "csv-parse/sync";

// Target measurement IDs that differ between implementations
const DIVERGENT_IDS = [
  "726b441f-eb43-47d9-8f3c-845d164e5a5b", // TS accepts, PY rejects
  "1a98b2c3-e023-4757-8d01-d35ef2fb363e", // PY accepts, TS rejects
  "510977fa-9d3f-4b50-a667-e676a0cc0791", // PY accepts, TS rejects
  "70d7918e-87d9-4968-84a4-b2bfec488e76", // PY accepts, TS rejects
  "86233705-0332-44f1-bc69-bc796220f598", // PY accepts, TS rejects
];

// Import TypeScript implementation
import { loadConfig } from "./typescript_lib/src/weight-processor-lib/core/config";
import { InMemoryStore } from "./typescript_lib/src/index";
import { WeightProcessorService } from "./services/weight_processor_service";

interface CsvRow {
  id?: string;
  user_id: string;
  value_quantity?: string;
  unit: string;
  timestamp?: string;
  effective_date_time?: string;
  effectiveDateTime?: string;
  source_type: string;
  [key: string]: any;
}

function parseTimestamp(dateStr: string): Date {
  if (!dateStr) return new Date();

  try {
    if (dateStr.includes("T")) {
      const normalized = dateStr.replace("Z", "+00:00");
      return new Date(normalized);
    } else if (dateStr.includes(" ")) {
      return new Date(dateStr.replace(" ", "T") + "Z");
    } else {
      return new Date(dateStr + "T00:00:00Z");
    }
  } catch {
    return new Date();
  }
}

async function main() {
  console.log("=== Debugging Divergent Measurements ===\n");
  console.log("Target IDs:");
  for (const id of DIVERGENT_IDS) {
    console.log(`  ${id}`);
  }
  console.log();

  // Load CSV
  const csvPath = "test_user.csv";
  const content = readFileSync(csvPath, "utf-8");
  const records = parse(content, {
    columns: true,
    skip_empty_lines: true,
    trim: true,
  }) as CsvRow[];

  // Extract all measurements for the user
  const userMeasurements: any[] = [];
  let divergentMeasurements: any[] = [];

  for (const row of records) {
    const measurementId = row.id || row.measurement_id;
    const userId = row.user_id;
    const weightStr = (row.value_quantity || row.weight || "").trim();

    if (!userId || !measurementId || !weightStr) continue;

    const weight = parseFloat(weightStr);
    if (isNaN(weight) || weight <= 0) continue;

    const dateStr = row.effective_date_time || row.effectiveDateTime || row.timestamp || "";
    const source = row.source_type || "unknown";
    const unit = (row.unit || "").trim();

    if (!unit) continue;

    const timestamp = parseTimestamp(dateStr);

    const measurement = {
      measurementId,
      weight,
      unit,
      timestamp,
      source,
    };

    userMeasurements.push(measurement);

    if (DIVERGENT_IDS.includes(measurementId)) {
      divergentMeasurements.push({ ...measurement, originalRow: row });
    }
  }

  console.log(`Loaded ${userMeasurements.length} measurements for user`);
  console.log(`Found ${divergentMeasurements.length} divergent measurements\n`);

  // Sort by timestamp
  userMeasurements.sort((a, b) => a.timestamp.getTime() - b.timestamp.getTime());

  // Initialize TypeScript implementation
  const config = loadConfig();
  const store = new InMemoryStore();
  const service = new WeightProcessorService(store, config);

  const userId = "ADC64C0B-CB46-41F9-BDA0-CC11A35942D7";

  // Process all measurements
  console.log("Processing all measurements...\n");
  const response = await service.processBatch(userId, userMeasurements);

  console.log(`Total processed: ${response.measurements_processed}`);
  console.log(`Total accepted: ${response.measurements_accepted}`);
  console.log(`Total rejected: ${response.measurements_rejected}\n`);

  // Analyze divergent measurements
  console.log("=== Divergent Measurement Analysis ===\n");

  for (const divergent of divergentMeasurements) {
    // Find the result for this measurement
    const measurementIndex = userMeasurements.findIndex(
      m => m.measurementId === divergent.measurementId
    );

    if (measurementIndex >= 0 && measurementIndex < response.results.length) {
      const result = response.results[measurementIndex];

      console.log(`ID: ${divergent.measurementId}`);
      console.log(`  Timestamp: ${divergent.timestamp.toISOString()}`);
      console.log(`  Source: ${divergent.source}`);
      console.log(`  Weight: ${divergent.weight} ${divergent.unit}`);
      console.log(`  Accepted: ${result.accepted}`);
      console.log(`  Quality Score: ${result.quality_score?.toFixed(6)}`);

      if (result.quality_components) {
        console.log(`  Quality Components:`);
        for (const [component, score] of Object.entries(result.quality_components)) {
          console.log(`    ${component}: ${(score as number).toFixed(6)}`);
        }
      }

      if (!result.accepted) {
        console.log(`  Rejection Reason: ${result.rejection_reason || 'Unknown'}`);
      }

      console.log(`  Kalman Estimate: ${result.kalman_estimate?.toFixed(3)}`);
      console.log(`  Innovation: ${result.innovation?.toFixed(3)}`);
      console.log(`  Normalized Innovation: ${result.normalized_innovation?.toFixed(3)}`);
      console.log();
    }
  }
}

main().catch(console.error);
